### DISCUSSION TITLE
How Generate Python stubs for FreeCAD C++

### DISCUSSION THREAD

**Discussion by oursland** [FreeCAD/FreeCAD#16405] [Code contribution discussions]
This is an approach to generating Python stubs for IDE autocompletion ("IntelliSense").

1. Build FreeCAD.  In my instance, this is built from source using the `conda` package manager for dependencies on macOS.  
2. Install within the FreeCAD environment the package [`pybind11-stubgen`](https://github.com/sizmailov/pybind11-stubgen.git).  This means that if you run `FreeCADCmd -c` then from within the console you should be able to run `import stubgen` without error.
3. Copy the attached [stubgen.py.txt](https://github.com/user-attachments/files/16939571/stubgen.py.txt) to your FreeCAD source directory.  This is a modified version of [\_\_init\_\_.py from pybind11_stubgen](https://github.com/sizmailov/pybind11-stubgen/blob/v2.4.2/pybind11_stubgen/__init__.py) that has been horribly hacked to parse the `FreeCAD.{so,dylib,dll}` library for `pybind11`-generated symbols.
4. Run the command `FreeCADCmd stubgen.py --pass -o stubs` to generate the stubs in the `stubs/` directory.  The attached 
[stubs/FreeCAD.pyi.txt](https://github.com/user-attachments/files/16939594/FreeCAD.pyi.txt) is the stubs file generated from my use of this process.

Then load in the stubs using your IDE's provisions.  For VS Code, I added to `.vscode/settings.json` the line `"python.analysis.stubPath": "${workspaceFolder}/stubs",`.

* the attached files have had `.txt` appended to the filename to get past GitHub's file type filter.

**Comment by bgbsww** [comment]
That's " ...  `import pybind11-stubgen` without error." above.  stubgen alone doesn't exist :)

In CLion, if you simply take the newly created `stubs` directory in the tree, right click it, select Mark Directory As -> Library Files the stubs will start working in the IDE.

**Comment by bgbsww** [comment]
My suggestion for this is to add the `stubs/` directory and the `stubs/FreeCAD.pyi` file to the repository since it won't change much and having an obsolete version is better than having no version.  We can also add the `stubgen.py` file, perhaps in `tools/stubs` along with a README.

Then the Dev guide can reference this, and or automagic CI tools to rebuild the stubs file can be created.

**Reply by oursland** [reply]
@looooo and I have been working a bit towards a `pixi` integration.  [`pixi`](https://pixi.sh/latest/) is a bit like `npm` for Python, enabling the management of [environments](https://pixi.sh/latest/features/environment/), [managing lock files](https://pixi.sh/latest/features/lockfile/), and [run-scripts/tasks](https://pixi.sh/latest/features/advanced_tasks/).

What do you think about including python stubs generation as a run script?

We may also go the extra step to auto generate and commit the bindings, but I'd like to get a little more consensus before having the CI system generate commits in the source tree.

The work for `pixi` is on @looooo's PR here: https://github.com/FreeCAD/FreeCAD/pull/16795

**Reply by bgbsww** [reply]
Good point.  The CI might be a step too far.  Pixi seems interesting and adding a run script make sense to me.

**Reply by furgo16** [reply]
This is all great, thanks. If I may share my thoughts, I would favor @bgbsww's approach as a first MVP that is simple and gets anyone 80% (or more) along the way there. Stub generation as a run script with pixi could be the next step.

My reluctance is simply that if the script needs to be run by the developer manually (as opposed to CI), it's yet another new technology to learn, which, as any, is also not free of quirks, bugs and integration issues with multiple platforms.
